### PR TITLE
New features for sass_layer; rename before/after

### PR DESCRIPTION
**Pull request by jcheng5**
(For discussion after the holidays)

**WARNING: BREAKING CHANGES**

- Adds ability to add file attachments and "tags" to sass_layer
- Renames before/after to defaults/rules, and adds a "declarations" between them
- REMOVES the ability for htmlDependencies to be on arbitrary sass input objects, now they'll only be respected when they're on actual sass layers

### CODE REVIEW

File: R/layers.R [hunk 1]
```diff
@@ -6,41 +6,68 @@
...
 #' # However, by stacking sass_layer()s, we have ability to place
 #' # SASS both before and after some other sass (e.g., core)
 #' # Here we place a red default _before_ the blue default and export the
 #' # color SASS variable as a CSS variable _after_ the core
-#' red_layer <- sass_layer(red, after = ":root{ --color: #{$color}; }")
+#' red_layer <- sass_layer(red, rules = ":root{ --color: #{$color}; }")
 #' sass(sass_layer_merge(core, red_layer))
 #' sass(sass_layer_merge(core, red_layer, sass_layer(green)))
 #'
+#' @name sass_layer
```
Comment (cpsievert) [inline R/layers.R]: Should this be `@rdname`?

File: R/layers.R [hunk 2]
```diff
@@ -6,41 +6,68 @@
...
 sass_layer_merge <- function(...) {
   layers <- dropNulls(rlang::list2(...))
   is_layer <- vapply(layers, is_sass_layer, logical(1))
   layers[!is_layer] <- lapply(layers[!is_layer], function(x) {
-    sass_layer(after = x)
+    sass_layer(rules = x)
```
Comment (cpsievert) [inline R/layers.R]: Just so I remember, we should document this behavior.

File: R/layers.R [hunk 3]
```diff
@@ -50,50 +77,126 @@ sass_layer <- function(before = "", after = "", html_deps = NULL) {
...
-  # i.e., htmltools::htmlDependencies()
-  html_deps <- attr(x, "html_dependencies", TRUE)
-  if (inherits(html_deps, "html_dependency")) {
-    html_deps <- list(html_deps)
+  dest <- names(attachments)
+  if (is.null(dest)) {
+    dest <- rep_len("", length(attachments))
   }
-  deps <- c(deps, html_deps)
+  src <- attachments
```
Comment (cpsievert) [inline R/layers.R]:
```suggestion
src <- as.character(attachments)
```

File: R/sass.R [hunk 4]
```diff
@@ -56,17 +68,8 @@ sass <- function(input = NULL, options = sass_options(), output = NULL,
...
-        }, warning = function(w) {
-          stop(conditionMessage(w))
-        })
-        return(invisible())
-      } else {
-        # TODO: Set encoding to UTF-8?
-        css <- paste(readLines(cache_file), collapse = "\n")
-      }
+      # TODO: Set encoding to UTF-8?
+      css <- paste(readLines(cache_file), collapse = "\n")
```
Comment (cpsievert) [inline R/sass.R]: Why the changes here?

File: R/sass.R [hunk 5]
```diff
@@ -35,13 +40,20 @@
...
     stop("Please construct the compile options using `sass_options()`.")
   }
   if (!inherits(cache_options, "sass_cache_options")) {
     stop("Please construct the cache options using `sass_cache_options()`.")
   }
+  if (!is.null(output) && !dir.exists(fs::path_dir(output))) {
+    stop("The output directory '", fs::path_dir(output), "' does not exist")
+  }
+  if (is.null(output) && isTRUE(write_attachments)) {
+    stop("sass(write_attachments=TRUE) cannot be used when output=NULL")
```
Comment (cpsievert) [inline R/sass.R]: I wonder how difficult it would be to do something like base64 encoding in this case?

File: R/layers.R [hunk 6]
```diff
@@ -50,50 +77,126 @@ sass_layer <- function(before = "", after = "", html_deps = NULL) {
...
+  check_files <- function(files, ok, message) {
+    if (any(!ok)) {
+      stop(call. = FALSE, message, ": ",
+        paste0("'", files[!ok], "'", collapse = ", "))
+    }
   }
 
-  deps
+  check_files(dest, !fs::is_absolute_path(dest),
+    "File attachment names must be relative (destination) paths")
```
Comment (cpsievert) [inline R/layers.R]:
```suggestion
    "File attachment destination paths must be relative.")
```

File: R/layers.R [hunk 7]
```diff
@@ -50,50 +77,126 @@ sass_layer <- function(before = "", after = "", html_deps = NULL) {
...
+
+  check_files(src, fs::is_absolute_path(src),
+    "File attachment names must be absolute (source) paths")
```
Comment (cpsievert) [inline R/layers.R]:
```suggestion
    "File attachment source path(s) must be absolute path(s).")
```

File: tests/testthat/test-layers.R [hunk 8]
```diff
@@ -1,21 +1,169 @@
...
 blue <- list(color = "blue !default")
 red <- list(color = "red !default")
 green <- list(color = "green !default")
-core <- sass_layer(before = blue, after = "body { color: $color; }")
+core <- sass_layer(
+  defaults = blue,
+  declarations = "@function invert($color, $amount: 100%) {
+    $inverse: change-color($color, $hue: hue($color) + 180);
+    @return mix($inverse, $color, $amount);
+  }",
```
Comment (cpsievert) [inline tests/testthat/test-layers.R]: DYK sass has a built in `invert()` that appears to do the same? 

```
> sass::sass("@debug invert(blue)")
stdin:1 DEBUG: yellow
```

File: R/sass.R [hunk 9]
```diff
@@ -35,13 +40,20 @@
 #'   sass_file(tmp_file)
 #' ))
 sass <- function(input = NULL, options = sass_options(), output = NULL,
-  cache_options = sass_cache_options()) {
+  cache_options = sass_cache_options(), write_attachments = FALSE) {
```
Comment (cpsievert) [inline R/sass.R]: Maybe this should default to `write_attachments = "auto"` which would resolve to `TRUE` if layers are present

File: tests/testthat/test-layers.R (hunk 8 above)
Comment (jcheng5) [reply tests/testthat/test-layers.R]: Note to self: use a different name

File: R/sass.R (hunk 9 above)
Comment (jcheng5) [reply R/sass.R]: Compromise approach of `write_attachments = NULL` would detect the presence of attachments and warn.

File: R/layers.R (hunk 3 above)
Comment (jcheng5) [reply R/layers.R]: I used `unname` instead
